### PR TITLE
Add VectorNonlinearFunction

### PR DESCRIPTION
**Pull request by odow**
Resurrecting `VectorNonlinearFunction` from the depths of https://github.com/jump-dev/MathOptInterface.jl/pull/2059.

There are two main motivating problems for this:

 1. `VectorNonlinearFunction-in-Complements` constrains for nonlinear complementarity problems
     - https://github.com/chkwon/PATHSolver.jl/pull/82
 2. `ObjectiveFunction{VectorNonlinearFunction}` for nonlinear multi-objective problems
     - https://github.com/jump-dev/MultiObjectiveAlgorithms.jl/pull/68

Documentation

https://jump.dev/MathOptInterface.jl/previews/PR2201/reference/standard_form/#MathOptInterface.VectorNonlinearFunction

I also removed this from #2059 because there we a lot of issues with the bridges, but I have a better handle on what to do now, and we can easily run `solver-tests` to catch the culprits.

https://github.com/jump-dev/MathOptInterface.jl/actions/runs/5218217869

### CODE REVIEW

**Comment by odow** [comment]
Marking this as blocked by #2203

**Comment by odow** [comment]
This is an interesting suggestion. We'd need backward compat for `(::VectorOfVariables).variables`, but a generic `AbstractVectorFunction` that is a vector of scalar terms would be useful.

**Comment by odow** [comment]
This is not fun. I think I'm going to revert the `GenericVectorFunction` changes to `MOI.VectorOfVariables`. There are too many little issues that crop up. Perhaps we can revisit once this PR is merged, and once we have a solution for the `operate` mess.

**Comment by blegat** [comment]
> There are too many little issues that crop up.

The tests seem to pass except for some printing issues though

**Comment by odow** [comment]
> The tests seem to pass except for some printing issues though

The tests don't cover the code enough. If you try running the tests with Clp/SCS etc lots of things break.

**Comment by odow** [comment]
This is still blocked by the `operate` issue, but we're getting closer: https://github.com/jump-dev/MathOptInterface.jl/pull/2215

File: src/Bridges/Constraint/bridges/soc_rsoc.jl
Comment (odow) [inline src/Bridges/Constraint/bridges/soc_rsoc.jl:81]: Fixed separately in https://github.com/jump-dev/MathOptInterface.jl/pull/2216

**Comment by odow** [comment]
https://github.com/jump-dev/MathOptInterface.jl/actions/runs/5341127312

**Comment by odow** [comment]
> If you try running the tests with SCS etc lots of things break.

Okay. One massive source of confusion for me had been failures with SCS. Which were because the bridge in SCS was wrong: https://github.com/jump-dev/SCS.jl/pull/277.

It used to do this:
```julia
julia> MOI.Bridges.print_active_bridges(model, MOI.VectorNonlinearFunction, MOI.PositiveSemidefiniteConeTriangle)
 * Unsupported constraint: MOI.GenericVectorFunction{Any}-in-MOI.PositiveSemidefiniteConeTriangle
 |  bridged by:
 |   SCS.ScaledPSDConeBridge{Float64, G} where G
 |  may introduce:
 |   * Supported constraint: MOI.VectorAffineFunction{Float64}-in-SCS.ScaledPSDCone
```
So no wonder lots of weird unrelated things broke.

But I only dug deep enough once I figured out it wasn't a method error with the various operate methods.

**Comment by odow** [comment]
So I was looking into `VectorNonlinearFunction-in-PositiveSemidefiniteConeTriangle` etc. And there are a number of thorny issues:

 * `Utilities.eachscalar` assumes that each row of the vector function is `scalar_type(F)`. But the row of a VectorNonlinearFunction might not be ScalarNonlinearFunction.
 * Easy fix is to convert every row, but then what's the difference between `Utilities.scalarize` and `Utilities.eachscalar`?
 * Operations don't simplify. So `f = sin(x)` and then `-(f, f)` doesn't simplify to `0`. So we get an expression swelling problem that prevents `Utilities.isapprox_zero` from working. The main problem with this is in the `Bridges.Constraint.SquareBridge`, which checks if off-diagonals are equivalent by checking if their difference is zero.

**Comment by blegat** [comment]
> Easy fix is to convert every row, but then what's the difference between `Utilities.scalarize` and `Utilities.eachscalar`?

Yes, I think every row should be an SNF. The difference of `scalarize` is that it allocates the vector of functions while `eachscalar` does not. If you have a lot or rows but you're only going to get one, `eachscalar` will be faster.
In fact, `scalarize` is really like conversion to a `GenericVectorFunction`

> The main problem with this is in the `Bridges.Constraint.SquareBridge`, which checks if off-diagonals are equivalent by checking if their difference is zero. 

Maybe it should use `isequal_canonical` or something like that

**Comment by odow** [comment]
Should `scalarize` be a copy? Currently it isn't always:

https://github.com/jump-dev/MathOptInterface.jl/blob/5afc08bbd153e3dab3d7f5d60420a3bea6f3da19/src/Utilities/functions.jl#L2110-L2120

Ditto with `vectorize`

https://github.com/jump-dev/MathOptInterface.jl/blob/5afc08bbd153e3dab3d7f5d60420a3bea6f3da19/src/Utilities/functions.jl#L2044-L2057

**Comment by blegat** [comment]
I don't think it has to be a copy. `MA.operate(scalarize, f)` should be a copy :laughing:

**Comment by odow** [comment]
https://github.com/jump-dev/MathOptInterface.jl/actions/runs/5383997349

File: test/Bridges/Constraint/square.jl
Comment (odow) [inline test/Bridges/Constraint/square.jl:249]: @blegat this is now working.

But if you print `model` you'll get a horrible function, because VectorNonlinear doesn't simplify `(f(x) - g(x)) + g(x) --> f(x)` at the expression level.

**Comment by odow** [comment]
PATHSolver.jl and MultiObjectiveAlgorithms.jl are both passing tests with this branch, so I don't think we're far away. Just need https://github.com/jump-dev/MathOptInterface.jl/actions/runs/5383997349 to finish.

File: src/Bridges/Constraint/bridges/vectorize.jl
Comment (odow) [inline src/Bridges/Constraint/bridges/vectorize.jl:52]: So many assumptions. Here's a fun one: we assumed that you could call 
`convert(::AbstractVectorFunction, ::AbstractScalarFunction)`.

**Comment by odow** [comment]
https://github.com/jump-dev/MathOptInterface.jl/actions/runs/5408332786

File: src/functions.jl
Comment (odow) [inline src/functions.jl:1027]: So I wondered about introducing `:identity` as a special-case nonlinear argument.

Then `VectorNonlinearFunction` would be `Vector{ScalarNonlinearFunction}`, and there wouldn't be this distinction between `scalarize` returning the `Vector{Any}` and `eachscalar` returning an iterator that converted each element to `ScalarNonlinearFunction`.

I guess `identity` is the same as univariate `:+`, but it's more explicit as to what is happening.

Comment (blegat) [reply src/functions.jl:1027]: Yes, I also prefer `identity`. Indeed, I can see the issue that `scalar_type` is `ScalarNonlinearFunction` so `eachscalar` has to convert to `ScalarNonlinearFunction which is inconsistent. Maybe that's a good argument against `Vector{Any}`

Comment (odow) [reply src/functions.jl:1170]: So I've been experimenting with this. It really doesn't add anything, and only complicates stuff. let's just leave as `:+` for now.

**Comment by odow** [comment]
https://github.com/jump-dev/MathOptInterface.jl/actions/runs/5756841023

**Comment by odow** [comment]
https://github.com/jump-dev/MathOptInterface.jl/actions/runs/5778904348

File: src/Bridges/Constraint/bridges/scalarize.jl
Comment (odow) [inline src/Bridges/Constraint/bridges/scalarize.jl:266]:
```suggestion
   ::MOI.ConstraintFunction,
```

**Comment by odow** [comment]
@blegat @mlubin this could use a review.

File: src/Bridges/Constraint/bridges/scalarize.jl
Comment (odow) [inline src/Bridges/Constraint/bridges/scalarize.jl:266]:
```suggestion
    ::MOI.ConstraintFunction,
```

File: src/functions.jl
Comment (odow) [inline src/functions.jl:708]: Should we make this 
```suggestion
    rows::Vector{ScalarNonlinearFunction}
```

Comment (odow) [reply src/functions.jl:708]: @blegat, take a look at this as-is with `Any`. I was playing around with `ScalarNonlinearFunction`, and it just ends up requiring a bunch of `convert` in annoying places. If we force each row to be `+(arg)`, then the downstream solvers are still going to have to special case a bunch of stuff, and they'll still need support for `ScalarAffineFunction` etc, because they can be arguments to `ScalarNonlinearFunction`.

The only problem with the current `Any` is the subtlety around `eachscalar` and `scalarize`.

Comment (blegat) [reply src/functions.jl:708]: > I was playing around with `ScalarNonlinearFunction`, and it just ends up requiring a bunch of `convert` in annoying places

Could you be more explicit ? Naively, I wouldn't think it would so I would need some examples to convince me ^^

> If we force each row to be `+(arg)`, then the downstream solvers are still going to have to special case a bunch of stuff, and they'll still need support for `ScalarAffineFunction` etc, because they can be arguments to `ScalarNonlinearFunction`.

Not really if they use `MOI.Nonlinear.Model` for AD. If they use the expression graph, they are probably still going to parse the expression and do the obvious simplifications so this simplification might just come as a special case.
It's going to cause a bunch of troubles that accessing the fields, `scalarize`, `eachscalar` and `scalar_type` don't match. I don't see an added value of that design that would outweight the added complexity of having these inconsistencies.

File: src/Utilities/functions.jl
Comment (odow) [inline src/Utilities/functions.jl:822]:
```suggestion
    return it.f.rows[output_index]
```

File: src/functions.jl
Comment (odow) [reply src/functions.jl]: Okay, take a look at this now. I think I got it to a nice point. The automatic conversion is still a bit weird. But maybe that's okay.

Comment (odow) [reply src/functions.jl]: [PATHSolver](https://github.com/chkwon/PATHSolver.jl/pull/82/files#diff-64cebae681efceded214317cd18b678aaefed5658af7a123f9b01433313a2844R231-R238) now needs this

image

where previously it could just require that `xi = f.rows[i+N]::MOI.VariableIndex`.

Comment (blegat) [reply src/functions.jl]: It makes sense, it's similar and consistent with Indicator which transforms the `VariableIndex` for having it fit into the `VectorAffineFunction`. It would be nice to have tools to make it easier though.

File: src/Utilities/functions.jl
Comment (blegat) [reply src/Utilities/functions.jl:493]: For instance, having
```julia
function substitute_variables(map, f)
    r = substitute(map, f)
    if f isa MOI.VariableIndex && ...
    end
    return r
end
```
and then replace all implementations of `substitute_variables` to implementations for `substitute`

Comment (blegat) [reply src/Utilities/functions.jl:493]: Then you can use `substitute` instead of `_unstable_substitute_variables`

Comment (odow) [reply src/Utilities/functions.jl:493]: Yeah only to avoid the error in the substitute variables for a variable.

File: src/Utilities/functions.jl
Comment (blegat) [reply src/Utilities/functions.jl:553]: Yes

File: src/Utilities/functions.jl
Comment (blegat) [reply src/Utilities/functions.jl:493]: Here is a draft, it might be more explicit
https://github.com/jump-dev/MathOptInterface.jl/pull/2250

Comment (blegat) [reply src/Utilities/functions.jl:493]: Ah but then, the `substitute` is unstable for `VariableIndex`, I see what you mean now. But maybe it does not matter now that small unions are fast. I wrote that a while ago

Comment (odow) [reply src/Utilities/functions.jl]: I've simplified and removed `_unstable_substitute_variables`

**Review by blegat — APPROVED** [review]

**Comment by mlubin** [comment]
I'll take a look over the weekend

**Review by mlubin — APPROVED** [review]

**Pull request merged**
